Add handler to run execution in separate goroutine

linxiulei opened the pull request:
This handler allows running execution prior to actual serving in a separate goroutine when serving requests. Doing so benefits cases in serving long running requests because it allows freeing memory used by the separate goroutine and keeps the serving routines slim.

#### What type of PR is this?


/kind cleanup

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Fixes #120901

#### Special notes for your reviewer:

Observed 150 MiB memory reduced in stack memory in the test that had 5000 active watch requests. The heap memory stayed roughly the same.

#### Does this PR introduce a user-facing change?

```release-note
An optimization is implemented to reduce stack memory usage for watch requests.  It is can be disabled with the feature gate: APIServingWithRoutine=false
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

aojea commented:
/cc

alexzielenski commented:
Looks promising! Thank you!
/triage accepted

linxiulei commented:
/retest

linxiulei commented:
/cc @wojtek-t @liggitt

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -65,8 +66,6 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
 // serveWatch will serve a watch response.
 // TODO: the functionality in this method and in WatchServer.Serve is not cleanly decoupled.
 func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions negotiation.MediaTypeOptions, req *http.Request, w http.ResponseWriter, timeout time.Duration, metricsScope string) {
-	defer watcher.Stop()
```
What if we return somewhere in the middle of this func (on error)? Then the watch will not be stopped and will effectively leak...

linxiulei replied:
Done despite the change is a bit verbose

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -218,89 +218,113 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-		case event, ok := <-ch:
-			if !ok {
-				// End of results.
+	serveFunc := func() {
```
I would benefit the PR to split it into smaller steps in which at least the first one is completely no-op, just a refactoring. It would also help with reviewing it.

linxiulei replied:
Done

linxiulei commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -258,17 +258,12 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
 				timeout = time.Duration(float64(minRequestTimeout) * (rand.Float64() + 1.0))
 			}
 			klog.V(3).InfoS("Starting watch", "path", req.URL.Path, "resourceVersion", opts.ResourceVersion, "labels", opts.LabelSelector, "fields", opts.FieldSelector, "timeout", timeout)
-			ctx, cancel := context.WithTimeout(ctx, timeout)
```
I noticed we use [TimeoutFactory](https://github.com/kubernetes/kubernetes/blob/78f6266dc5406d0e835344da06bb5447227a1f0d/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L52)  so wondered if this is still needed

wojtek-t replied:
If this is some kind of cleanup, then let's make a separate PR for it that is juts cleaning this up.

wojtek-t replied:
I don't think we can just remove it - I think we should switch to just calling `defer cancel()` inside `serve` function.

linxiulei replied:
You're right. Fixed.

linxiulei committed:
Add handler to run watch serving in separate goroutine

This handler allows running execution prior to actual serving in a separate
goroutine when serving requests. Doing so benefits cases in serving long running
requests because it allows freeing memory used by the separate goroutine
and keeps the serving routines slim.

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,78 @@
+// freeing stack memory used in preparation Goroutine for better memory efficiency.
+func WithRoutine(handler http.Handler, longRunning request.LongRunningRequestCheck, resolver request.RequestInfoResolver) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestInfo, err := resolver.NewRequestInfo(req)
```
The reason I created RequestInfo was to put this handler wrap at as top level as possible to have clean stack (ie. haven't gone into too many calls) so that when we go back to this handler, the stack memory usage is low, which is the idea for the optimization

wojtek-t replied:
I'm not a fan of it - we shouldn't be duplicating the work and optimizing one thing at the cost of breaking something else that was optimized before.
This is also somewhat related to my comment about putting this handler in front of WithPanic...

linxiulei replied:
Okay, I moved the routine up in the chain to reuse the work. FTR, I observed difference of 2MiB in a quick test comparison for stack memory usage with 5k watch connections, which is completely acceptable given this PR will reduce from ~200MiB to ~60MiB.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -226,75 +226,79 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+				event.Object = &unknown
+				metrics.WatchEventsSizes.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Observe(float64(len(unknown.Raw)))
+
+				*outEvent = metav1.WatchEvent{}
```
Please rebase on top of https://github.com/kubernetes/kubernetes/pull/120300

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -298,13 +322,28 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	defer ws.Close()
 	done := make(chan struct{})
+	// ensure the connection times out
+	timeoutCh, cleanup := s.TimeoutFactory.TimeoutCh()
```
This change seems ortogonal to this PR - can you move it to a separate PR?

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -179,6 +186,7 @@ type WatchServer struct {
 	// used to encode the nested object in the watch stream
 	EmbeddedEncoder runtime.Encoder
 
+	MemoryAllocator      runtime.MemoryAllocator
```
I'm looking for ways to do this change more incrementally and this seems to be a good one - can you move it to a separate PR (together with the first commit)?

In general - the simpler the change is, the easier it is to get convinced that it's correct.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -968,6 +968,7 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithMuxAndDiscoveryComplete(handler, c.lifecycleSignals.MuxAndDiscoveryComplete.Signaled())
 	handler = genericfilters.WithPanicRecovery(handler, c.RequestInfoResolver)
 	handler = genericapifilters.WithAuditInit(handler)
+	handler = genericfilters.WithRoutine(handler, c.LongRunningFunc, c.RequestInfoResolver)
```
I understand the idea, but that seems risky to me.
In general `WithPanicRecovery` is supposed to protect us from panics, so by putting something earlier in the chain, we are risking that it will not play its role correctly...
@liggitt - FYI

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -298,13 +322,28 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	serveFunc()
+	serve := func() {
+		requestInfo, _ := apirequest.RequestInfoFrom(req.Context())
+		metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, func() {
```
Same here - moving this part to that separate PR I mentioned above would make it easier to reason about everything.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -258,17 +258,12 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
-			metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, func() {
-				serveWatch(watcher, scope, outputMediaType, req, w, timeout, metrics.CleanListScope(ctx, &opts))
-			})
+			serveWatch(watcher, scope, outputMediaType, req, w, timeout, metrics.CleanListScope(ctx, &opts))
```
Once https://github.com/kubernetes/kubernetes/pull/122036 merges, please rebase this PR - it should be waaaay simpler to follow now.

linxiulei replied:
Will do ASAP. Thanks very much!

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -270,11 +269,19 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
+				t.F = serve
+			} else {
+				serve()
+			}
```
This is now much easier to follow - thanks!

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -1035,6 +1035,7 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 		handler = genericapifilters.WithTracing(handler, c.TracerProvider)
 	}
 	handler = genericapifilters.WithLatencyTrackers(handler)
+	handler = genericfilters.WithRoutine(handler, c.LongRunningFunc)
```
Do you have any updated data if the gains from this PR changed after moving this here into the handler chain?

linxiulei replied:
Yes, 

```
# original
go_memstats_stack_inuse_bytes 2.0201472e+08

# after this PR
go_memstats_stack_sys_bytes 6.2816256e+07
```

wojtek-t replied:
Hmm - you're saying this is improving usage by 70%?
That sounds a bit suspicious to me...

linxiulei replied:
Just to clarify, the 70% is from the whole PR, not from moving the handler chain. Moving the position of the handler chain only impacts 2 MiB or so, which might even be just statistical bias.

To elaborate on the 70% (ie. the 140MiB reduction), golang simply doubles the stack size if current stack size is exceeded, and due to the lengthy call depth in watch serving (also most of other resource serving calls as well), the [stack](https://github.com/kubernetes/kubernetes/issues/117777#issuecomment-1723055664) results in 32k in stack memory size for the primary goroutine for each watch connection. This RP reduces the stack memory to 4k or 2k (initial size), so the estimated memory reduction is (32k - 4k) * watch_connections (5k) = 140 MiB, which is perfectly matched what I observed from tests.

linxiulei replied:
I don't think so. Please see diff and full metrics (noting, the total usage diff is lower than claimed because Golang holds free memory and doesn't return back to system for various reasons)

Diff

```
diff /tmp/before /tmp/after
3c3
< go_memstats_alloc_bytes 5.53660392e+08
---
> go_memstats_alloc_bytes 5.32985304e+08
6c6
< go_memstats_alloc_bytes_total 2.3088914e+09
---
> go_memstats_alloc_bytes_total 2.383644024e+09
9c9
< go_memstats_buck_hash_sys_bytes 1.775686e+06
---
> go_memstats_buck_hash_sys_bytes 1.80191e+06
12c12
< go_memstats_frees_total 7.569195e+06
---
> go_memstats_frees_total 8.519271e+06
15c15
< go_memstats_gc_sys_bytes 2.3851392e+07
---
> go_memstats_gc_sys_bytes 2.4460152e+07
18c18
< go_memstats_heap_alloc_bytes 5.53660392e+08
---
> go_memstats_heap_alloc_bytes 5.32985304e+08
21c21
< go_memstats_heap_idle_bytes 1.27090688e+08
---
> go_memstats_heap_idle_bytes 1.6523264e+08
24c24
< go_memstats_heap_inuse_bytes 5.99801856e+08
---
> go_memstats_heap_inuse_bytes 6.5036288e+08
27c27
< go_memstats_heap_objects 5.08416e+06
---
> go_memstats_heap_objects 4.75143e+06
30c30
< go_memstats_heap_released_bytes 3.784704e+06
---
> go_memstats_heap_released_bytes 2.531328e+06
33c33
< go_memstats_heap_sys_bytes 7.26892544e+08
---
> go_memstats_heap_sys_bytes 8.1559552e+08
36c36
< go_memstats_last_gc_time_seconds 1.7029097843228054e+09
---
> go_memstats_last_gc_time_seconds 1.702909472212391e+09
42c42
< go_memstats_mallocs_total 1.2653355e+07
---
> go_memstats_mallocs_total 1.3270701e+07
51c51
< go_memstats_mspan_inuse_bytes 1.1601912e+07
---
> go_memstats_mspan_inuse_bytes 1.2004608e+07
54c54
< go_memstats_mspan_sys_bytes 1.1993856e+07
---
> go_memstats_mspan_sys_bytes 1.3216056e+07
57c57
< go_memstats_next_gc_bytes 1.049976896e+09
---
> go_memstats_next_gc_bytes 1.035159256e+09
60c60
< go_memstats_other_sys_bytes 3.952434e+06
---
> go_memstats_other_sys_bytes 3.930258e+06
63c63
< go_memstats_stack_inuse_bytes 1.95854336e+08
---
> go_memstats_stack_inuse_bytes 6.1014016e+07
66c66
< go_memstats_stack_sys_bytes 1.95854336e+08
---
> go_memstats_stack_sys_bytes 6.1014016e+07
69c69
< go_memstats_sys_bytes 9.64351448e+08
---
> go_memstats_sys_bytes 9.20049112e+08
```

Before

```
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 5.53660392e+08
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 2.3088914e+09
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.775686e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 7.569195e+06
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 2.3851392e+07
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 5.53660392e+08
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.27090688e+08
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 5.99801856e+08
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 5.08416e+06
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 3.784704e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 7.26892544e+08
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.7029097843228054e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 1.2653355e+07
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 19200
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 31200
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 1.1601912e+07
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 1.1993856e+07
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.049976896e+09
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 3.952434e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 1.95854336e+08
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 1.95854336e+08
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 9.64351448e+08
```

After

```
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 5.32985304e+08
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 2.383644024e+09
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.80191e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 8.519271e+06
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 2.4460152e+07
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 5.32985304e+08
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.6523264e+08
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 6.5036288e+08
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 4.75143e+06
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 2.531328e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 8.1559552e+08
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.702909472212391e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 1.3270701e+07
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 19200
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 31200
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 1.2004608e+07
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 1.3216056e+07
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.035159256e+09
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 3.930258e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 6.1014016e+07
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 6.1014016e+07
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 9.20049112e+08
```

aojea replied:
please add a comment explaining this,
also I commented here https://github.com/kubernetes/kubernetes/pull/120902#issuecomment-1861795535 but due the complexity introduced I may suggest have a feature flag to control the feature at least during one release

linxiulei replied:
Done

linxiulei commented:
Rebased to the latest and addressed comments. @wojtek-t PTAL, thanks!

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -259,22 +260,32 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
+			serve := func() {
+				requestInfo, _ := request.RequestInfoFrom(ctx)
+				metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, func() {
+					defer cancel()
```
I would move it higher - to the `serve` func itself.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -259,22 +260,32 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
 			}
 			klog.V(3).InfoS("Starting watch", "path", req.URL.Path, "resourceVersion", opts.ResourceVersion, "labels", opts.LabelSelector, "fields", opts.FieldSelector, "timeout", timeout)
 			ctx, cancel := context.WithTimeout(ctx, timeout)
-			defer cancel()
```
is it possible for anything between here and the serve() invocation to panic? if so, we'll lose the cancel() call.

linxiulei replied:
That's possible, I add the defer back to call cancel if it returns early.

aojea replied:
I'm confused with this cancel, are we sure we need this complexity of returnEarly bool? context cancellation only propagates to childrens, no?

wojtek-t replied:
+1 - I don't really follow what returnEarly gives us here...

linxiulei replied:
If either of `rw.Watch` or `serveWatchHandler`, will the ctx with cancellation leak since we don't defer calling cancel()?

linxiulei replied:
Followed @wojtek-t 's suggestion to make it easier to understand

linxiulei commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,73 @@
+		req = req.WithContext(WithTask(ctx, &Task{}))
+
+		c := make(chan struct{})
+		go func() {
```
I pass the panic to parent goroutine, but not sure if it's the correct way to do it.

wojtek-t replied:
@liggitt - I think that the current implementation is ok, because:
1. this handler is put into handler chain after WithPanicHandler
2. so if the serve function panics, we catch the panic, and we panic ourselves, which in turn should be handled correctly by WithPanicHandler

So I think it works fine - but we definitely need a test proving this.

linxiulei replied:
Added tests to prove this.

aojea commented:
This seems to complex and hard to prevent any side effect, what do you think of rolling this out behind a temporary feature flag so we can get some signal and roll the change back safely if necessary?

aojea commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,79 @@
+
+		req = req.WithContext(WithTask(ctx, &Task{}))
+		var panicVar interface{}
+		c := make(chan struct{})
```
c := make(chan any, 1)

wojtek-t replied:
+1 for all Anthonio's comments around it

linxiulei replied:
Done

aojea commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,79 @@
+		}
+
+		req = req.WithContext(WithTask(ctx, &Task{}))
+		var panicVar interface{}
```
you don't need this, you can send it through the channel

aojea commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,79 @@
+		go func() {
+			defer func() {
+				if r := recover(); r != nil {
+					panicVar = r
```
panicCh <- r

aojea commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,79 @@
+			handler.ServeHTTP(w, req)
+		}()
+
+		<-c
+		if panicVar != nil {
+			panic(panicVar)
+		}
```
select {
	case p := <-panicChan:
		panic(p)
	case <-ctx.Done():
		return
	}

linxiulei replied:
I don't think we need to handle ctx.Done() here because handler in deeper chain should handle it and clean stuff. If we return here, the clean function in the deeper chain is never executed to clean resource up.

aojea commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,79 @@
+// WithRoutine returns an http.Handler that executes preparation of long running requests (i.e. watches)
+// in a separate Goroutine and then serves the long running request in the main Goroutine. Doing so allows
+// freeing stack memory used in preparation Goroutine for better memory efficiency.
+func WithRoutine(handler http.Handler, longRunning request.LongRunningRequestCheck) http.Handler {
```
can we add an unit test to this?

wojtek-t replied:
We definitely need it - I was first trying to understand all the potential consequences of this change...

aojea commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -270,11 +276,21 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
+					handler.ServeHTTP(w, req)
+				})
+			}
+			t := genericfilters.TaskFrom(req.Context())
+			if t != nil {
+				t.F = serve
+			} else {
+				serve()
+			}
```
this requires a comment explaining the magic here

linxiulei commented:
Added unit tests, feature gate and comments. Updated release notes to mention feature gate.

linxiulei commented:
/retest

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,77 @@
+const taskKey taskKeyType = iota
+
+type Task struct {
+	F func()
```
Maybe call it Handler? Or Func?

Let's use more meaningful names.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/server/filters/routine.go:
```diff
@@ -0,0 +1,77 @@
+			handler.ServeHTTP(w, req)
+		}()
+
+		if p := <-panicCh; p != nil {
```
```
p, ok := <- panicCh
if !ok {
  panic(p)
}
```

wojtek-t reviewed (commented):
I want to look deeper into tests, but I think it's going the right direction.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -1038,6 +1038,11 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithLatencyTrackers(handler)
+	// WithRoutine will execute future handlers in a separate goroutine and serving
+	// handler in current goroutine to minimize the stack memory usage.
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.APIServingWithRoutine) {
```
Please also add a comment that this has to be put after WithPanicRecovery (as it is now), but to ensure that someone will not change it later.

Also please squash the commits.

linxiulei replied:
Added comment and squashed.

wojtek-t reviewed (commented):
One last comment - other than that LGTM.

wojtek-t commented:
/lgtm
/approve
